### ISSUE TITLE
End skill from Proactive message

### ISSUE THREAD

**Issue by tomSauret847** [microsoft/botframework-solutions#3626]
#### What project is affected?
Virtual Assistant and skill
#### What language is this in?
TypeScript
#### What happens?
We are unable to end the skill and have it send the EoC to the VA from a proactive message
#### What are the steps to reproduce this issue?
Generate the VA and Skill,
Implement the proactive messaging middle-ware
send a message to the skill and have the skill end and send the EoC to the VA
#### What were you expecting to happen?
We are interfacing a skill with our call center chat system
this requires running a synchronous loop to retrieve messages from the chat system
As the SDK 4 framework does not allow synchronous loops to be ran we are using proactive messages to return the messages from the chat system
When our agents end the chat we will need to be able to end the skill and return the customer back to the VA from the proactive message
#### Can you share any logs, error output, etc.?
We are receiving the "Cannot preform "get" on a proxy that has been revoked" error from the synchronous loop

**Comment by Batta32** [comment]
Hi @tomSauret847. Thanks for reporting this issue. We will be reviewing it and we will back to you soon!

**Comment by peterinnesmsft** [comment]
We are scheduling further investigation of this issue for next week, where we will aim to have a repro and root cause analysis completed by end of next week.

**Comment by tomSauret847** [comment]
@peterinnesmsft If needed I can be available for a offline call to fully explain what the skill is going to be used for and what we are trying to achieve with this particular skill.

**Comment by Batta32** [comment]
@tomSauret847 - we are currently working on this issue so we came up with some validations that we want to ask you

1. Do you have the Virtual Assistant and Skill both in TypeScript?
1. Are you using the [ProactiveStateMiddleware](https://github.com/microsoft/botframework-solutions/blob/master/sdk/typescript/libraries/bot-solutions/src/proactive/proactiveStateMiddleware.ts) for the proactive messaging middleware that you mentioned?
1. Have you configured the middleware in the Virtual Assistant only?
1. Is the middleware correctly executed when the Skill ends?
1. Are you using the Bot Framework Emulator to check this issue?

**Comment by tomSauret847** [comment]
@Batta32 to answer your questions
1. Yes both VA and skill is in Typescript
2. No we are not using the ProactiveStateMiddleware
3. No neither skill or VA
5. Yes we are using the emulator for testing locally and will be using direct line and Teams channels in production

Use case. 
We are interfacing the skill with a internal chat system that requires polling loop to check it every 2 seconds. Since synchronous are not allowed in the bot framework we are needing to send the messages and events (ending event) back through a proactive message. Currently I am using the following function to send the messages back
await this.defaultAdapter.continueConversation(convoId as Partial<ConversationReference>, async turnContext => {
      // we can send the Live Agent's message here by continuing the conversation
      await turnContext.sendActivity('message from chat server);

  }); 

This does not allow any access to the Turn State to be able to call the "endDialog()" to end the skill and return control back to the VA. If we keep these functions in the DefaultActivityHandler we can get the functionality working but the VA will throw an error invoking the skill as we are not starting any dialogs. If you need more context please let me know.

**Comment by Batta32** [comment]
Thanks @tomSauret847 - we will let you know if we need more context 😊.

We identified this [sample](https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/typescript_nodejs/16.proactive-messages) of BotBuilder-Samples which appears to have the same code to send the messages back as you commented in the [index.ts](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/16.proactive-messages/src/index.ts#L90).

As soon as we have any update, we will back to you!

**Comment by Batta32** [comment]
Hi @tomSauret847, we need you to give us a full overview of what you're doing in order to figure out the scenario and reproduce your issue. Here are some initial **questions** we are particulary interested in:

1. Where are you adding the function to send the messages back? We think that you are adding this as an endpoint in the Virtual Assistant, is this correct?
    ```typescript
    await this.defaultAdapter.continueConversation(convoId as Partial, async turnContext => {
    // we can send the Live Agent's message here by continuing the conversation
    await turnContext.sendActivity('message from chat server);
    });
    ```
1. Is the above function the entire code that you added to have the proactive message?
1. How the proactive message is executed? Is this executed by an event? Which one?
1. What do you mean with:
    > If we keep these functions in the DefaultActivityHandler we can get the functionality working but the VA will throw an error invoking the skill as we are not starting any dialogs.
1. Can you confirm if the proactive message is used to send the "End Of Conversation" to the Skill?

Also, we analyzed the attached [sample](https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/typescript_nodejs/16.proactive-messages) and noticed that the [api/notify](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/16.proactive-messages/src/index.ts#L105) endpoint listens for incoming messages and sends the proactive messages to the users. This is manually executed when the endpoint receives a request with the notifications.

![image](https://user-images.githubusercontent.com/11904023/91875330-b9af4900-ec51-11ea-8650-4c29873b081f.png)

Finally, we reviewed the following documents:
- [Send proactive notifications to users](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=csharp)
- [BotBuilder-Samples/16.proactive-messages](https://github.com/microsoft/BotBuilder-Samples/blob/main/samples/typescript_nodejs/16.proactive-messages)

We will be attentive to your answer 😊.

**Comment by tomSauret847** [comment]
@Batta32 thank you again for looking into this. I have attached the DefaultActivityHandler where we are able to get the functions to work but the VA will throw the error that the Skill could not be invoked. I have also attached the MainDialog showing where we are running a continuous loop that will poll a on premise server as part of the chat interaction. We do have this working in our current SDK 3 bot but SDK 3 allows for the synchronous and asynchronous to be used. Let me know if you need more information. 
[defaultActivityHandler.txt](https://github.com/microsoft/botframework-solutions/files/5158008/defaultActivityHandler.txt)
[mainDialog.txt](https://github.com/microsoft/botframework-solutions/files/5158011/mainDialog.txt)

**Comment by Batta32** [comment]
Thanks again @tomSauret847, we will be reviewing the attached documents and we will back to you later 😊.

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, before we move forward trying to reproduce the issue on our side, we want to make sure that we correctly understand your scenario and problem, along with the architecture where you are implementing the bots.

When we started to review this issue we found similarities with your previous issue, [#3437](https://github.com/microsoft/botframework-solutions/issues/3437), regarding how your bot architecture is implemented. Using the information in both issues, we created what we think what would be the **workflow** of the system:

1.  The user initiates the conversation with the Virtual Assistant
2.  At some point, the user may ask to talk with a Live Agent
3.  The Virtual Assistant executes the Skill that transfer the user to the Live Agent
4.  The Virtual Assistant also forwards all the conversation messages to the Skill so the Live Agent is able to follow the conversation that the user had
5.  The Skill receives the conversation from the Virtual Assistant and sends them to the Live Agent's chat
6.  The Skill starts to poll the user's chat every 2 seconds and sending them to the Live Agent's chat
7.  After the Live Agent finishes talking with the user, the Skills sends an EndOfConversation to the Virtual Assistant
8.  After ending the Skill, the user is transferred back to the Virtual Assistant

Based on the previous workflow, we have some **questions**:

1.  **Step 6**.
    1.  You also mentioned having issues on the step 6, as if you await the polling method the skill will never start because since it enters a loop to check for messages, not allowing the logic that starts the skill to be executed. But if you remove the await, you get the *Cannot perform get* issue. This seems an expected behaviour as per [microsoft/botbuilder-js#606](https://github.com/Microsoft/botbuilder-js/issues/606).
    2.  We are not sure if the skill is polling the user's or the agent's chat messages, but we think that it is the user's chat. Can you confirm us this?
3.  **Step 7**. We think that this is where you are having the main issue, the skill sends the EndOfConversation to the Virtual Assistant, but the Virtual Assistant never receives it so the user is never transferred from the agent to the assistant. Is this correct?
4.  Would it be possible for you to share with us a branch where we can easily reproduce this issue? We had trouble trying to replicate the behaviour.

_Analyzed enviroment_
![image](https://user-images.githubusercontent.com/39467613/92415118-dd7cfe00-f12d-11ea-980a-510deb314f86.png)

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk for looking into this. Sorry for the late reply, the answers to your questions are,
Step 6. We are trying to poll the agent chat servers, not the user messages. Below is a flow diagram showing the process we are attempting to complete. 
![image](https://user-images.githubusercontent.com/61023857/92508021-87f82e00-f1cd-11ea-99f2-a814def52322.png)
The highlighted step in the synchronous loop that is causing the need to use proactive messages to return the agent messages to the user. Would we need to move this to another app to remove it from the skill?

Step 7. Being able to end the skill and return control back to the VA from a proactive message is the reason for this issue. The VA is not receiving the EoC. The proactive messages are working while interfacing with the skill but they are not working when connected to the skill through the VA. We are using the /api/notify endpoint in the skill not the VA

The code I have shared is most of what we have completed. The API calls that are referenced in the MainDialog.ts are reaching to the chat server and returning back the array of events that the "pollChat" function is processing.

**Comment by Batta32** [comment]
Thanks @tomSauret847 - we will be reviewing the information detailed above and we will back to you as soon as we have any update!

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, thank you for your detailed diagram, it really helped us to understand better the situation. We made some tests, and we believe that we were able to **reproduce** the behaviour that you are observing on your side. Could you please **confirm** it?

We [replicated the logic](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L78-L90) of your polling loop on a Skill Sample. It simulates an agent ending the conversation while it is looping, based on the mainDialog code [you shared](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-685040193) with us.

To reproduce this behaviour, we prepared the following **repro steps**:

**Steps:**

1.  Clone the branch [feature/southworks/issue-3626/polling-loop](https://github.com/southworks/botframework-solutions/tree/feature/southworks/issue-3626/polling-loop)
2.  Open a console and navigate to the [TypeScript Skill Sample](https://github.com/southworks/botframework-solutions/tree/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill)
3.  Run `npm install` to install the dependencies
4.  Run `npm build` to compile the solution
5.  Deploy the Skill Sample
6.  Start the Skill
7.  Start the sample dialog by sending `run sample dialog`
8.  When the skill asks for your name, reply with `issue` to trigger the dummy poll method
9.  The skill will start to generate random numbers, once the skill draws a 4, it will try to stop and end the dialog
10. After trying to end the dialog, the skill will keep looping ever 2 seconds, and the EoC will not be sent,

When the dummy method calls to end the dialog, it keeps looping, tracing every 2 seconds, and never returns to the Virtual Assistant. You can find a gif of this behaviour at the end of the comment.

## More Details


_The behaviour of the dummy polling method with the never ending loop_
![NeverEndingLoop](https://user-images.githubusercontent.com/39467613/92782572-70f73e80-f37b-11ea-87f8-0c97d5f031ff.gif)

**Comment by VictorGrycuk** [comment]
@tomSauret847 - we modified the dummy method to include a possible workaround.

The same skill sample has a [modified dummy poll](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L92-L109) chat method with a possible workaround. To test this follow these steps:

**Steps**
1.  If you haven't followed the [previous repro steps](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-690610885), follow them up to **step 7**
2.  Instead of replying with `issue`, reply with `workaround`to trigger the dummy poll method
3.  Again, the skill will start to generate random numbers
4.  Once it draws a 4, it will stop the conversation and return a EndOfConversation

We think that this might be related to how the loop method is implemented, so we modified the logic of determine if we should loop/poll again *before* looping or polling.

Since the polling chat is recursive, it calls itself every 2 seconds, we pass a parameter to determine if it should call itself again or not. If not, the next time it will just return before it has a chance to call itself. Check [our implementation](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L92-L109), and see if you can implement something simiar on yor side.

You can find a gif of this behaviour at the end of the comment.

We will be attentive to your reply 😊.

## More Details
*The behaviour of the dummy polling method after applying a workaround*
![EndingLoop](https://user-images.githubusercontent.com/39467613/92786565-b9642b80-f37e-11ea-9188-289fb9416069.gif)

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk for this. Yes we are able to run the lop and end the skill with the work around. But when a message is typed in while in the loop it starts another instance of the loop. This will cause problems as it will increase the polling with every message the customer types in. Do you know of a way that when we are in the loop to intercept the customer's messages to prevent starting a new instance of the loop? See the below

![image](https://user-images.githubusercontent.com/61023857/92796267-a8172100-f376-11ea-9312-066a8daa22a5.png)

**Comment by VictorGrycuk** [comment]
Thanks @tomSauret847, we will look into this and get back to you as soon as we have an update.

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk I did do some testing and found if I put a check in the "greetUser" step it will not restart the loop. I added the following code into that step. Let me know you thoughts on this possible fix.

        if (userProfile.chatting) {
            // send customers message to the chat servers
            return SampleDialog.EndOfTurn;
        }
        // we start the chat and get our participant ID
        const start = await this.chatStart(stepContext, userProfile.ininGroup, userProfile.attempts);
        userProfile.chatting = true;
        userProfile.participantId = start;
        // we are not making a turn so save user state
        await this.userState.saveChanges(stepContext.context, false);

**Comment by tomSauret847** [comment]
@VictorGrycuk I have been doing some more testing and the VA is still trowing a error 2 minutes into the Skill running. It appears as though the message is not getting received by the skill until the loop stops running. Even looking at the gif from above the last message you send before the loop starts does not send until the loop completes. this will cause the below error to be thrown
Error: socket hang up
    at createHangUpError (_http_client.js:323:15)
    at Socket.socketOnEnd (_http_client.js:426:23)
    at Socket.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)

Here is a screen shot of the emulator chat showing the message that was sent and 2 minutes later the error that was thrown

![image](https://user-images.githubusercontent.com/61023857/92974621-45905480-f44c-11ea-8e2a-e9ed686be016.png)

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, we analyzed the [mentioned](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-690733083) possible fix. While it **does make sense** to validate if the loop is running, and if it does, prevent it to run it again, **we suggest** to decouple the logic of the loop and the logic that handles the messages.

As an example, we modified the skillSample and added a new dialog called [loopDialog](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/loopDialog.ts) that the only logic it has is the loop and the `sendActivity`, which in your case would be the webchat messages.

The [sampleDialog](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts) constructor was modified to [initialize a loopDialog](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L54), and the [greetUser](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L65-L81) method was also modified to start or end the loop depending on the message that sends the user. You can check this in the attached gif.

The loop has to run in parallel of the logic that handles the messages, because once the loop is called, it wont return control due to the `await`. We did several tests trying to separate these logics, and we think it would be a good idea to keep the logic of the loop in a separate dialog than the one dialog that handles the messages.

Regarding the hang up error, we will research about this. We will be attentive to your comments. 🙂 

*The Virtual Assistant running the loopDialog through the Skill*
![LoopDialog 1](https://user-images.githubusercontent.com/39467613/93384436-35f18100-f83b-11ea-942e-dfe06048d5d5.gif)

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk I have checked out the sample you provided and tested it. I have found it restarts the MainDialog over again when the customer types in a message. I will ensure I decouple the loop from the logic and wait for a response on why we are getting the socket hang up.

**Comment by Batta32** [comment]
Hi @tomSauret847 - we **reproduced** the "_socket hang up error_", both using a Virtual Assistant and running the Skill alone.
We believe this is happening because the flow of the code is stopped at the await that calls the loop, and the communication time out in [restify](https://www.npmjs.com/package/restify) is set to **2 minutes by default**.

Repro steps
1. Clone the branch [feature/southworks/issue-3626/polling-loop](https://github.com/southworks/botframework-solutions/tree/feature/southworks/issue-3626/polling-loop)
1. Open a console and navigate to the [TypeScript Skill Sample](https://github.com/southworks/botframework-solutions/tree/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill)
1. Run `npm install` to install the dependencies
1. Run `npm build` to compile the solution
1. Deploy the Skill Sample
1. Start the Skill
1. Start the sample dialog by sending `run sample dialog`
1. When the skill asks for your name, reply with `start looping` to trigger the dummy poll method
1. Wait two minutes
1. An error should appear mentioning the socket hang up error

We will be doing tests regarding the socket hang up error message, trying to find a workaround or methods to mitigate it.

Here you can find some related documents:
- restify/node-restify#798
- [NodeJS - What does “socket hang up” actually mean?](https://stackoverflow.com/questions/16995184/nodejs-what-does-socket-hang-up-actually-mean)
- [How to increase the socket timeout on the server side using Restify?](https://stackoverflow.com/questions/15305068/how-to-increase-the-socket-timeout-on-the-server-side-using-restify)
- [How to debug a socket hang up error in NodeJS](https://stackoverflow.com/questions/10814481/how-to-debug-a-socket-hang-up-error-in-nodejs/11542134#11542134)
- [Override 2 minute timeout in Restify](https://coder1.com/articles/override-2-minute-timeout-restify)
- microsoft/botbuilder-js#1705

_The hang up error reproduced using a Skill_
![image](https://user-images.githubusercontent.com/11904023/93614478-93efa700-f9a8-11ea-9902-4d5d202e75a7.png)

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, we further analyzed the issues you experienced and came to some **conclusions**.

### Main Problem

Like you [mentioned](https://github.com/microsoft/botframework-solutions/issues/3626#issue-681058804), we believe that the **main problem** was how the loop was implemented. Because you are awaiting a loop, the logic of the bot is stuck at the `await` until the loop ends. This not only made the handling of the messages difficult, if not impossible, but also generated the socket hang up error.

### Our Suggestion

-   **Implement** proactive messages. We did a [small spike](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/loopDialog.ts#L47-L68)  that implement simple proactive messages by storing the [adapter and conversation reference](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/loopDialog.ts#L42-L43) before starting the loop, and then using them to send the proactive messages without blocking the functionality of the skill.\
    However, please note that this does not work when the skill called form a virtual assistant. The loop will continue to work in the background, but the messages wont be sent to the virtual assistant. To make it work, you will have to do a proper proactive message between the loopDialog and the virtual assistant.
-   **Replace**  `setTimeout` with [setInterval](https://nodejs.org/api/timers.html#timers_setinterval_callback_delay_args), as this makes it much easier to handle the loop than using recursive calls.

### Mentioned errors

1.  [**Socket Hang Up Error**](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-691321349):
     1.  *Reason*: As explained in [last comment,](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-694930284) this is related to the default timeout of the [restify](https://www.npmjs.com/package/restify) communication, and we suspected this was happening because the logic of the skill is stuck at the `await` that calls the `pollChat` method.
     2.  *Fix*: It was needed to execute the looping logic without blocking the flow of the skill, and that meant to remove the `await` and run the loop in parallel.\
    We fixed this by replacing `setTimeout` with [setInterval](https://nodejs.org/api/timers.html#timers_setinterval_callback_delay_args), making the method that calls it [synchronous](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/loopDialog.ts#L47), and making the method that sends the activity [asynchronous](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/loopDialog.ts#L70-L74). That way, the asynchronous calls are encapsulated inside the `setInterval` and not in the bot logic.

2.  [**Duplicated Loops**](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-690688314):
     1.  *Reason*: logic of the loop was being instantiated every time that the loopDialog was called.
     2.  *Fix*: add a static property `isLooping` inside the loopDialog, that we set to true when the loop start, and then only start the loop if it is [not looping](https://github.com/southworks/botframework-solutions/blob/feature/southworks/issue-3626/polling-loop/templates/typescript/samples/sample-skill/src/dialogs/sampleDialog.ts#L73-L78).

3.  [**Sending Proactive EoC**](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-683102530):
     1.  *Reason*: as suggested above, our spike works correctly while the skill is active and in conversation, however, when it has to send activities after the control was sent to the virtual assistant.
     2.  *Fix*: one possible solution would be, using our approach as base, to send the messages from inside the loop to the virtual assistant proactively using direct line. That way, the loop will keep running in the background and sending the messages to the virtual assistant through direct line.

If you wish to implement proactive messages, here are some resources:

-   [Stack Overflow thread on which we based our proactive implementation](https://stackoverflow.com/a/62267157/10872646)
-   [016.proactive-messages @ BotBuilder-Samples](https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/typescript_nodejs/16.proactive-messages)
-   [Proactive Messaging @ botbuilder-proactivemessaging](https://github.com/lucashuet93/botbuilder-proactivemessaging)
     -   Also check his [node implementation](https://github.com/lucashuet93/botbuilder-proactivemessaging/tree/master/node)

Hope this helps to you! 🙂

*The loopDialog sending activities proactively in the background even after the sampleDialog is done.*
![proactive loop](https://user-images.githubusercontent.com/39467613/93928499-cd018180-fcf0-11ea-9dc2-b9603a632fb0.gif)

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk for the suggestions. I have found a couple issues with the implementation so far.

1. Using a static variable in the LoopingDialog will prevent multiple users from using the skill at the same time. This will lock the skill down so it can only be used by 1 customer at any one time.
2. By removing the await and letting the skill end control we are unable to capture the customers messages and forward them over to a live agent. The skill terminates and returns control back to the VA where it will route the conversation over to a new intent whenever the customer types in a message. This will make this implementation not feasible since we require the 2 way communication between the customer and the live agent. We do need to keep the skill active and intercept all of the customers messages and forward them to a API

It looks for our use case the best approach will be to keep the skill in running the loop and override the socket timeout to keep the conversation going. This is the solution that will allow us to intercept the customers messages and end the skill when the loop gets terminated by the agent leaving the conversation. Please let me know if you have any other suggestions on this issue as I would prefer not to override the socket timeout for this skill.

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, final **suggestions** below:

1.  Yes, you are right. We simply used a `static` property for the sake of simplicity. You will have to convert it into a `public` instance property of the LoopDialog.
2.  Awaiting on the loop wouldn't prevent the skill to communicate back to the user? We think that the loop that retrieves the messages from the web chat server should run independently of the logic of the bot. Instead of running off of a skill, it might be run off of the Virtual Assistant at the moment it executes the transfer.

@peterinnesmsft - since this is more of an architectural concern and not something related to the current solutions of the repository, maybe someone of your team can give us a better approach regarding this?

@tomSauret847 - let us know if we can help in something else specific of the solutions of the repository 😊.

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk Yes the skill is able to communicate back to the user, but we cannot receive the user's messages in the skill. We NEED those messages to forward to the chat server so the live agent can get responses from the user. By holding open the socket it will keep the skill active and allow us to get the user's messages to forward to the live agent. If we allow the skill to return back to the VA then we will no longer have access to the user's messages and they are unable to communicate with the live agent.

From your suggestion above the best route forward would be to create a app that will handle the looping logic and send the messages back to the VA through proactive messaging and remove the skill?

**Comment by VictorGrycuk** [comment]
@tomSauret847 - as far as we researched, the **best approach** that our team found is to have the loop logic independently of `AgenTransferSkill` logic, and have it send the messages proactively to the Virtual Assistant since it would not block the flow of the bot as with the `await`. This loop would be started by the AgentTransferSkill when it's called by the Virtual Assistant and stopped by the AgentTranserSkill before sending the EndOfConversation.

@peterinnesmsft - can your team give us their insights on what they think it would be the best approach for this issue?

Let us know if you want us to help you with something else 😊.

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk and @Batta32 for your assistance on this issue. We are exploring another possible solution that we are getting ready to test now. We have developed this skill in C# that will allow us to perform the loop required. We will be leaving our VA and other skills in TypeScript and just converting this skill that needs the loop to run while it is active.

**Comment by Batta32** [comment]
Great @tomSauret847! If you agree we can close this issue, but feel free to raise a new issue and we will help you 😊!

**Comment by tomSauret847** [comment]
Thank you again for your assistance @Batta32 I will close this issue and watch for the publish of the bot-solutions 1.0.1
